File: app/src/main/java/com/thesunnahrevival/sunnahassistant/viewmodels/SettingsViewModel.java
package com.thesunnahrevival.sunnahassistant.viewmodels;

import android.app.Application;
import android.content.Context;
import android.net.ConnectivityManager;

import com.thesunnahrevival.sunnahassistant.R;
import com.thesunnahrevival.sunnahassistant.data.AppSettings;
import com.thesunnahrevival.sunnahassistant.utilities.GeneralSaveDataAsyncTask;

import android.os.Handler;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

public class SettingsViewModel extends SunnahAssistantViewModel {

    public MutableLiveData<Boolean> closeWelcomeScreen = new MutableLiveData<>();
    public int mCounter = 1;
    private MutableLiveData<String> offsetValue = new MutableLiveData<>();
    private MutableLiveData<String> hijriChangeErrorMessage = new MutableLiveData<>();
    private MutableLiveData<String> updatePrayerSettingsMessage = new MutableLiveData<>();
    private int mOffsetSavedValue;


    public SettingsViewModel(@NonNull Application application) {
        super(application);
    }

    public void setOffsetSavedValue(int value) {
        mOffsetSavedValue = value;
        offsetValue.setValue(String.valueOf(value));
    }

    public LiveData<String> getOffSet() {
        return offsetValue;
    }

    public LiveData<String> getHijriChangeErrorMessage() {
        return hijriChangeErrorMessage;
    }

    public void updateHijriDate(int value) {
        if (value < -30 || value > 30)
            hijriChangeErrorMessage.setValue("Error! Value should be between -30 and 30");

        else if (isDeviceOffline(getApplication()))
            hijriChangeErrorMessage.setValue("Error Updating. Please Check your internet connection.");

        else if (value != mOffsetSavedValue) {
            hijriChangeErrorMessage.setValue("Updating...");
            mRepository.deleteHijriDate();
            if (mSettings.getValue() != null) {
                mSettings.getValue().setHijriOffSet(value);
                updateSettings();
                hijriChangeErrorMessage.setValue(null);
            }
            mOffsetSavedValue = value;
        } else
            hijriChangeErrorMessage.setValue("No changes made");
    }


    public LiveData<String> getCityName() {
        return Transformations.map(mSettings, AppSettings::getFormattedAddress);
    }

    public LiveData<String> getLocationUpdateMessage() {
        return Transformations.map(mRepository.getGeocodingData(), message -> {
            AppSettings settings = mSettings.getValue();
            if (message != null && !message.getResults().isEmpty() && settings != null) {
                float latitude = message.getResults().get(0).getGeometry().getLocation().getLat();
                float longitude = message.getResults().get(0).getGeometry().getLocation().getLng();
                settings.setLatitude(latitude);
                settings.setLongitude(longitude);
                settings.setFormattedAddress(message.getResults().get(0).getFormattedAddress());
                savePrayerTimesSettings();
                return "Successful:" + message.getResults().get(0).getFormattedAddress();
            }
            return null;
        });
    }

    public LiveData<String> getPrayerUpdateMessage() {
        return updatePrayerSettingsMessage;
    }

    public void setPrayerSettingsManually() {
        if (mSettings.getValue() != null) {
            mSettings.getValue().setAutomatic(false);
            mSettings.getValue().setFormattedAddress("");
            boolean shouldCloseWelcomeScreen = false;
            if (mSettings.getValue().isFirstLaunch()) {
                mSettings.getValue().setFirstLaunch(false);
                shouldCloseWelcomeScreen = true;
            }
            mSettings.getValue().setMonth(0);
            updateSettings();
            mRepository.deletePrayerTimesData();
            updatePrayerSettingsMessage.setValue(getApplication().getString(R.string.successfully_updated));
            if (shouldCloseWelcomeScreen)
                closeWelcomeScreen.setValue(true);
        }
    }

    private void savePrayerTimesSettings() {
        if (mCounter == 0 && !isDeviceOffline(getApplication())) {
            mCounter++;
            mRepository.deletePrayerTimesData();
            updatePrayerSettingsMessage.setValue("Updating...");
            Runnable runnable = () -> {
                AppSettings settings = mSettings.getValue();
                if (GeneralSaveDataAsyncTask.prayerDeleteTaskComplete && settings != null) {
                    settings.setAutomatic(true);
                    settings.setMonth(0); //Reset month to allow refreshing data
                    boolean shouldCloseWelcomeScreen = false;
                    if (mSettings.getValue().isFirstLaunch()) {
                        mSettings.getValue().setFirstLaunch(false);
                        shouldCloseWelcomeScreen = true;
                    }
                    updateSettings();
                    updatePrayerSettingsMessage.setValue("Successfully Updated.");
                    if (shouldCloseWelcomeScreen)
                        closeWelcomeScreen.setValue(true);
                }
            };
            new Handler().postDelayed(runnable, 2000);
        } else if (isDeviceOffline(getApplication())) {
            updatePrayerSettingsMessage.setValue(getApplication().getString(R.string.no_internet_connection));
        } else if (mCounter != 0)
            updatePrayerSettingsMessage.setValue(getApplication().getString(R.string.no_changes));


    }

    /**
     * Checks to see if the location input has changed by comparing it with
     * what is stored in settings
     */
    private boolean validateLocationChanged(String address) {
        return mSettings.getValue() != null && !mSettings.getValue().getFormattedAddress().matches(address);
    }

    public void initiateFetchingGeocodingData(String address) {
        if (isDeviceOffline(getApplication()))
            updatePrayerSettingsMessage.setValue(getApplication().getString(R.string.error_updating_location));
        else if (TextUtils.isEmpty(address))
            updatePrayerSettingsMessage.setValue("Location cannot be empty");
        else if (validateLocationChanged(address)) {
            mCounter = 0;
            mRepository.fetchGeocodingData(address);
            updatePrayerSettingsMessage.setValue("Updating...");
        } else
            savePrayerTimesSettings();
    }

    private boolean isDeviceOffline(Context context) {
        ConnectivityManager connectivityManager = (
                (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE));

        return connectivityManager.getActiveNetworkInfo() == null ||
                !connectivityManager.getActiveNetworkInfo().isConnected();
    }

    public void updateSettings() {
        mRepository.updateAppSettings(mSettings.getValue());
    }

}

File: app/src/main/java/com/thesunnahrevival/sunnahassistant/viewmodels/SettingsViewModel.kt
package com.thesunnahrevival.sunnahassistant.viewmodels

import android.app.Application
import android.content.Context
import android.net.ConnectivityManager
import android.net.Uri
import android.widget.Toast
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import com.thesunnahrevival.sunnahassistant.data.SunnahAssistantRepository
import com.thesunnahrevival.sunnahassistant.data.model.AppSettings
import com.thesunnahrevival.sunnahassistant.data.model.GeocodingData
import com.thesunnahrevival.sunnahassistant.utilities.SunnahAssistantUtil
import com.thesunnahrevival.sunnahassistant.utilities.TimeDateUtil

class SettingsViewModel(application: Application) : AndroidViewModel(application) {
    private val mRepository: SunnahAssistantRepository = SunnahAssistantRepository.getInstance(application)
    val settings: LiveData<AppSettings>
    var isSettingsUpdated: Boolean private set
    var month = TimeDateUtil.getMonthNumber(System.currentTimeMillis())

    private val settingsValue: AppSettings?
        get() = settings.value

    init {
        settings = mRepository.appSettings
        isSettingsUpdated = false
    }

    fun updateDisplayHijriDateSettings(isDisplay: Boolean) {
        settingsValue?.let {
            it.isDisplayHijriDate = isDisplay
            updateSettings()
        }

    }


    fun updateAutomaticPrayerTimeSettings(isEnabled: Boolean) {
        settingsValue?.let {
            if (it.isAutomatic != isEnabled) {
                it.isAutomatic = isEnabled
                updateSettings()
            }
        }

    }

    fun fetchGeocodingData(address: String?) {
        mRepository.fetchGeocodingData(address)
    }

    fun updateGeneratedPrayerTimes() {
        settingsValue?.let {
            if (isLoadFreshData) {
                if (it.isAutomatic) {
                    mRepository.updateGeneratedPrayerTimes(
                            it.latitude, it.longitude, it.toString(), TimeDateUtil.getYear(System.currentTimeMillis()),
                            it.method, it.asrCalculationMethod, it.latitudeAdjustmentMethod)
                    //Save the Month in User Settings to prevent re-fetching the data the current month
                    updateSavedMonth()
                    Toast.makeText(getApplication(), "Refreshing", Toast.LENGTH_LONG).show()
                }
            }
        }
    }

    private fun updateSavedMonth() {
        settingsValue?.let{
            if (it.month != month) {
                it.month = month
                updateSettings()
            }
        }

    }

    fun updateSavedSpinnerPosition(position: Int) {
        settingsValue?.let {
            if (it.savedSpinnerPosition != position) {
                it.savedSpinnerPosition = position
                updateSettings()
            }
        }
    }

    fun updateIsShowOnBoardingTutorial(isShowOnBoardingTutorial: Boolean) {
        settingsValue?.let {
            if (it.isShowOnBoardingTutorial != isShowOnBoardingTutorial) {
                it.isShowOnBoardingTutorial = isShowOnBoardingTutorial
                updateSettings()
            }
        }
    }

    fun observeGeocodingApiData(): LiveData<GeocodingData> {
        return mRepository.geocodingData
    }

    fun updateLocationDetails(data: GeocodingData) {
        settingsValue?.let {
            val result = data.results[0]
            it.formattedAddress = result.formattedAddress
            it.latitude = result.geometry.location.lat
            it.longitude = result.geometry.location.lng
            updateSettings()

        }
    }

    fun updateCalculationMethod(calculationMethod: Int) {
        settingsValue?.let {
            if (it.method != calculationMethod) {
                it.method = calculationMethod
                updateSettings()
            }
        }

    }

    fun updateAsrCalculationMethod(calculationMethod: Int) {
        settingsValue?.let {
            if (it.asrCalculationMethod != calculationMethod) {
                it.asrCalculationMethod = calculationMethod
                updateSettings()
            }
        }
    }

    fun updateHigherLatitudeMethod(method: Int) {
        settingsValue?.let {
            if (it.latitudeAdjustmentMethod != method) {
                it.latitudeAdjustmentMethod = method
                updateSettings()
            }
        }
    }

    fun updatePrayerTimesData() {
        val settings = settingsValue
        if (isSettingsUpdated && settings != null && !settings.formattedAddress.matches("Location cannot be empty".toRegex())) {
            mRepository.deletePrayerTimesData()
            if (settings.isAutomatic) {
                settings.categories.remove(SunnahAssistantUtil.PRAYER)
                mRepository.generatePrayerTimes(
                        settings.latitude, settings.longitude, settings.month.toString(),
                        TimeDateUtil.getYear(System.currentTimeMillis()), settings.method, settings.asrCalculationMethod,
                        settings.latitudeAdjustmentMethod
                )
            }
            else
                settings.categories.add(SunnahAssistantUtil.PRAYER)
            updateCategories(settings.categories)
        }
    }

    fun updateLayout(isExpandedLayout: Boolean) {
        settingsValue?.let {
            if (it.isExpandedLayout != isExpandedLayout) {
                it.isExpandedLayout = isExpandedLayout
                updateSettings()
            }
        }
    }

    fun updateTheme(isLightMode: Boolean) {
        settingsValue?.let {
            if (it.isLightMode != isLightMode) {
                it.isLightMode = isLightMode
                updateSettings()
            }
        }
    }

    fun updateCategories(categories: HashSet<String>) {
        settingsValue?.let {
            it.categories = categories
            updateSettings()
        }
    }

    fun updatedDeletedCategories(deletedCategories: ArrayList<String>) {
        if (deletedCategories.isNotEmpty()) {
            mRepository.updateDeletedCategories(deletedCategories)
        }
    }

    fun updateNotificationSettings(notificationUri: Uri?, isVibrate: Boolean, priority: Int) {
        mRepository.updateNotificationSettings(notificationUri, isVibrate, priority)
    }

    fun updateRingtone(uri: Uri?) {
        settingsValue?.let {
           it.notificationToneUri = uri
            updateSettings()
        }
    }

    fun updateVibrationSettings(isVibrate: Boolean) {
        settingsValue?.let {
            it.isVibrate = isVibrate
            updateSettings()
        }
    }

    fun updateNotificationPriority(priority: Int) {
       settingsValue?.let {
           it.priority = priority
           updateSettings()
       }


    }

    fun showStickyNotification(isShow: Boolean) {
        settingsValue?.let {
            if (it.isShowNextReminderNotification != isShow) {
                it.isShowNextReminderNotification = isShow
                updateSettings()
            }
        }
    }

    fun updateIsFirstLaunch(isFirstLaunch :Boolean){
        settingsValue?.let {
            it.isFirstLaunch = isFirstLaunch
            updateSettings()
        }
    }

    fun updateIsAfterUpdate(isAfterUpdate :Boolean){
        settingsValue?.let {
            it.isAfterUpdate = isAfterUpdate
            updateSettings()
        }
    }

    private fun updateSettings() {
        isSettingsUpdated = true
        settingsValue?.let { mRepository.updateAppSettings(it) }
    }

    val isDeviceOffline: Boolean
        get() {
            val connectivityManager = getApplication<Application>().applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            return connectivityManager.activeNetworkInfo == null ||
                    !connectivityManager.activeNetworkInfo.isConnected
        }

    private val isLoadFreshData: Boolean
        get() = if (settingsValue != null)
            settingsValue?.month != month
        else true


}